### ISSUE TITLE
[v0.0.5.0] install-sshd.ps1: SDDL string contains an invalid sid or a sid that cannot be translated

### ISSUE THREAD

**Issue by pvdhoef** [PowerShell/Win32-OpenSSH#446]
When trying to install v0.0.5.0 with: 
```shell
powershell -executionpolicy bypass -file install-sshd.ps1
```

I get the following error:
```powershell
[SC] SetServiceObjectSecurity SUCCESS
[SC] ChangeServiceConfig SUCCESS
 Granting SeAssignPrimaryTokenPrivilege to NT SERVICE\SSHD   ... successful
Exception calling "SetSecurityDescriptorSddlForm" with "1" argument(s): "The SDDL string contains an invalid sid or a sid that cannot be translated.
Parameter name: sddlForm"
At C:\Program Files\OpenSSH-Win64\install-sshd.ps1:40 char:1
+ $acl.SetSecurityDescriptorSddlForm($sddl)
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : ArgumentException
```

The offending line in `install-sshd.ps1`:
```powershell
﻿$acl.SetSecurityDescriptorSddlForm($sddl)
```

**Comment by manojampalam** [comment]
What OS are you seeing this on? and what's your Powershell version?

As a work around, you may explicitly grant RW permissions to the following account
NT SERVICE\sshd 
on "logs" folder

**Comment by friism** [comment]
@manojampalam I see this too on Windows Server 2016 RTM.

```
PS C:\Program Files\OpenSSH-Win64> $PSVersionTable.PSVersion

Major  Minor  Build  Revision
-----  -----  -----  --------
5      1      14393  206
```

**Comment by jbaumbach** [comment]
Same issue here.  Windows Server 2016 Datacenter v1607 running in Azure.

```PS C:\openssh> $PSVersionTable.PSVersion

Major  Minor  Build  Revision
-----  -----  -----  --------
5      1      14393  576
```

**Comment by jbaumbach** [comment]
@manojampalam 

> As a work around, you may explicitly grant RW permissions to the following account
NT SERVICE\sshd

BTW, there's no "NT SERVICE" user on the system.  I assumed you meant "SERVICE", which does exist.

**Comment by friism** [comment]
@manojampalam yeah, do you have a scriptable workaround?

**Comment by pvdhoef** [comment]
Hi @manojampalam, the error occurs on a Windows 10 Enterprise box with PowerShell version ﻿5.1.14393.0.

**Comment by jbaumbach** [comment]
@manojampalam The service doesn't appear to be starting regardless of what permissions I apply to the logs folder.  This is in the event log:

`The application-specific permission settings do not grant Local Activation permission for the COM Server application with CLSID 
{D63B10C5-BB46-4990-A94F-E40B9D520160}
 and APPID 
{9CA88EE3-ACB7-47C8-AFC4-AB702511C276}
 to the user NT SERVICE\sshd SID (S-1-5-80-3847866527-469524349-687026318-516638107-1125189541) from address LocalHost (Using LRPC) running in the application container Unavailable SID (Unavailable). This security permission can be modified using the Component Services administrative tool.`

**Comment by manojampalam** [comment]
Can you please dump the output of the following:

$AdObj = New-Object System.Security.Principal.NTAccount('NT SERVICE\sshd')
$AdObj.Translate([System.Security.Principal.SecurityIdentifier]).Value

**Comment by manojampalam** [comment]
@jbaumbach were you able to work with earlier versions?

**Comment by friism** [comment]
@manojampalam 

```
PS C:\Program Files\OpenSSH-Win64> $AdObj = New-Object System.Security.Principal.NTAccount('NT SERVICE\sshd')
>> $AdObj.Translate([System.Security.Principal.SecurityIdentifier]).Value
>>
>>
S-1-5-80-3847866527-469524349-687026318-516638107-1125189541
PS C:\Program Files\OpenSSH-Win64>
```

I'm upgrading from `v0.0.3.0` which worked great.

**Comment by friism** [comment]
I sent a PR: https://github.com/PowerShell/openssh-portable/pull/34

**Comment by manojampalam** [comment]
Thanks @friism. I'll test it out and update the release payload Tue morning PST.

**Comment by manojampalam** [comment]
@friism, I've manually merged your PR and updated the payload on https://github.com/PowerShell/Win32-OpenSSH/releases/tag/v0.0.5.0.

Let me know if this fixed the issue.

**Comment by friism** [comment]
@manojampalam it appears to work, thanks!

I have a question: It looks like you just took the changes in my commit, created a new commit with you as author and pushed that to your branch. This eradicates any trace of me authoring the change. Instead, you could have just cherry-picked the commit into your branch. Is there any reason you did that, eg. that I had to sign a contribution agreement or something, and it was too much work for me to sign it?

**Comment by pvdhoef** [comment]
Installation works now without errors, thanks a lot for that!

Log files are now correctly written to `C:\Program Files\OpenSSH-Win64\logs`.

However, when trying to `ssh` to the machine I get:
```
ssh_exchange_identification: Connection closed by remote host
```
And in `logs\sshd`:
```
2408 11:42:39 421 Server listening on :: port 22.
2408 11:42:39 421 Server listening on 0.0.0.0 port 22.
2408 11:42:39 421 error: Couldn't create pid file "./sshd.pid": Permission denied
2408 11:43:34 999 error: fork: No such file or directory
```
Tried to set r/w permissions on `C:\Program Files\OpenSSH-Win64` for `NT SERVICE\sshd` but that didn't help.

After stopping the `sshd` service and starting it from the command line with `sshd.exe -d` I was able to `ssh` to the machine:
```
debug1: Bind to port 22 on ::.
Server listening on :: port 22.
debug1: socket:532, io:000001FC6FC8F970, fd:4
debug1: Bind to port 22 on 0.0.0.0.
Server listening on 0.0.0.0 port 22.
debug1: socket:536, io:000001FC6FC91100, fd:5
debug1: pipe - r-h:540,io:000001FC6FC42530,fd:6  w-h:544,io:000001FC6FC95500,fd:7
debug1: Server will not fork when running in debugging mode.
... etc
```
The message `Server will not fork when running in debugging mode` explains maybe why the `error: fork: No such file or directory` did not occur this time.

Hope this helps. If you need more information, please let me know.

**Comment by jbaumbach** [comment]
@manojampalam I'll be honest, I realized I was trying to fit a square peg in a round hole after mucking around in the registry, and just stood up an Ubuntu instance.  SFTP up and running in 5 minutes.   :(

**Comment by manojampalam** [comment]
@friism, apologies for the credits part. I'm working on multiple branches with little relation between them and took in your change as part of ongoing manual merges. Add credits line now in latestw_all branch.

@pvdhoef please open a new issue with the details you posted

@jbaumbach looks like you what you are hitting is unrelated to the install failure. Please open a  new issue.

**Comment by friism** [comment]
Now that I'm through the install, I'm seeing the same error @jbaumbach is:

```
The application-specific permission settings do not grant Local Activation permission for the COM Server application with CLSID 
{D63B10C5-BB46-4990-A94F-E40B9D520160}
 and APPID 
{9CA88EE3-ACB7-47C8-AFC4-AB702511C276}
 to the user NT SERVICE\sshd SID (S-1-5-80-3847866527-469524349-687026318-516638107-1125189541) from address LocalHost (Using LRPC) running in the application container Unavailable SID (Unavailable). This security permission can be modified using the Component Services administrative tool.
```

**Comment by friism** [comment]
I think it's this problem: https://support.microsoft.com/en-us/kb/899965

**Comment by manojampalam** [comment]
Can you see if setting sshd to run as SYSTEM gets you unblocked?
sc.exe config sshd obj= LocalSystem

**Comment by manojampalam** [comment]
@pvdhoef  looks like recent changes in "paths" related logic caused a regression. To work around, please install at a location without any spaces. Ex C:\openssh

**Comment by pvdhoef** [comment]
@manojampalam, I created a separate issue (#451) as requested. It can indeed be fixed by installing into a location without white space characters in the directory path it as you suggested.
I saw you already created a issue ( #450) for that.
